### PR TITLE
🐛Fix blocking call in the garbage collector

### LINKED ISSUE #2399: webserver blocked by unidentified reason
The two errors refer to the same function

```python
  _     ._   __/__   _ _  _  _ _/_   Recorded: 05:21:10  Samples:  1,
 /_//_/// /_\ / //_// / //_'/ //     Duration: 0.703     CPU time: 0.702,
/   _/                      v3.4.2,
,
Program: /home/scu/.venv/bin/simcore-service-webserver --config server-docker-prod.yaml,
,
0.654 instrumented  servicelib/monitor_slow_callbacks.py:23,
└─ 0.654 instrumented  aiodebug/log_slow_callbacks.py:16,
   └─ 0.654 _run  asyncio/events.py:143,
      └─ 0.654 collect_garbage_periodically  simcore_service_webserver/resource_manager/garbage_collector.py:63,
         └─ 0.654 collect_garbage  simcore_service_webserver/resource_manager/garbage_collector.py:89,
            └─ 0.654 remove_orphaned_services  simcore_service_webserver/resource_manager/garbage_collector.py:317,
               └─ 0.654 is_node_id_present_in_any_project_workbench  simcore_service_webserver/projects/projects_api.py:396,
                  └─ 0.654 get_all_node_ids_from_workbenches  simcore_service_webserver/projects/projects_db.py:788,
                     └─ 0.654 __anext__  aiopg/sa/result.py:343,
                        └─ 0.654 fetchone  aiopg/sa/result.py:377,
```

```python
   _     ._   __/__   _ _  _  _ _/_   Recorded: 05:24:49  Samples:  1,
  /_//_/// /_\ / //_// / //_'/ //     Duration: 0.664     CPU time: 0.665,
 /   _/                      v3.4.2,
 ,
 Program: /home/scu/.venv/bin/simcore-service-webserver --config server-docker-prod.yaml,
 ,
 0.588 instrumented  servicelib/monitor_slow_callbacks.py:23,
 └─ 0.588 instrumented  aiodebug/log_slow_callbacks.py:16,
    └─ 0.588 _run  asyncio/events.py:143,
       └─ 0.588 collect_garbage_periodically  simcore_service_webserver/resource_manager/garbage_collector.py:63,
          └─ 0.588 collect_garbage  simcore_service_webserver/resource_manager/garbage_collector.py:89,
             └─ 0.588 remove_orphaned_services  simcore_service_webserver/resource_manager/garbage_collector.py:317,
                └─ 0.588 is_node_id_present_in_any_project_workbench  simcore_service_webserver/projects/projects_api.py:396,
                   └─ 0.588 get_all_node_ids_from_workbenches  simcore_service_webserver/projects/projects_db.py:788,
                      └─ 0.588 __iter__  _collections_abc.py:760,
```

Looking at the function's code I was unable to identify what happened.

https://github.com/ITISFoundation/osparc-simcore/blob/bb62cf8b3b5c9228e2ee8e3a501499c0e39ea5a6/services/web/server/src/simcore_service_webserver/projects/projects_db.py#L788-L807

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

This PR optimizes a call to the database used by the **G**arbage **C**ollector that is asking:
- for all running services, find the one that are not in the database

To this end it was calling a SQL that would return all the nodeID in all the projects in the table. In master alone this would return >7000 entries. Times the number of running services.
Then this would be passed into a python Set, which is a blocking call twice.

This PR adds a function ```node_id_exists``` that runs a pSQL call that checks in the workbench field whether a specific nodeID exists. The computation happens on the postgres server and returns a yes/no type entry.

```BEFORE: 4.5s to call on about >1000 projects with 23-1434 nodes each```
```AFTER: 0.87s```

Bonus: 2 new tests

## Related issue/s
fixes #2399

### CODE REVIEW

**Review by pcrespov — APPROVED** [review]
great improvement! :tada: Added some comments below ... 
Q: how did you measure the time improvement?

File: services/web/server/src/simcore_service_webserver/projects/projects_db.py [hunk 1]
```diff
@@ -808,26 +808,28 @@ async def _get_tags_by_project(conn: SAConnection, project_id: str) -> List:
...
+                .select_from(projects)
+                .where(projects.c.workbench.op("->>")(f"{node_id}") != None)
+            )
+        assert num_entries is not None  # nosec
+        return bool(num_entries > 0)
 
-            return result
+    async def get_node_ids_from_project(self, project_uuid: str) -> Set[str]:
+        """Returns a set containing all the node_ids from project with project_uuid"""
+        result = set()
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_db.py:824]: well I spent already too long trying to make it work with sqlalchemy code but failed. so I resorted to the old one.

File: services/web/server/tests/unit/with_dbs/05/test_project_db.py [hunk 2]
```diff
@@ -670,3 +666,84 @@ async def test_patch_user_project_workbench_concurrently(
         creation_date=to_datetime(new_project["creationDate"]),
         last_change_date=latest_change_date,
     )
+
+
+@pytest.fixture()
+async def load_of_projects(
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/web/server/tests/unit/with_dbs/05/test_project_db.py:672]: I don't mind. ;)

File: services/web/server/tests/unit/with_dbs/05/test_project_db.py [hunk 3]
```diff
@@ -670,3 +666,84 @@ async def test_patch_user_project_workbench_concurrently(
...
+        *[
+            db_api.delete_user_project(logged_user["id"], f"{p_uuid}")
+            for p_uuid in all_created_projects
+        ]
+    )
+
+
+@pytest.mark.parametrize(
+    "user_role",
+    [UserRole.USER],
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/web/server/tests/unit/with_dbs/05/test_project_db.py:717]: you mean by putting it on the client fixture? but what if I do not use the logged_user ? I guess it's not a very big overhead...

File: services/web/server/src/simcore_service_webserver/projects/projects_db.py [hunk 4]
```diff
@@ -808,26 +808,28 @@ async def _get_tags_by_project(conn: SAConnection, project_id: str) -> List:
...
+        """Returns True if the node id exists in any of the available projects"""
         async with self.engine.acquire() as conn:
-            result = set()
-            query_result = await conn.execute(query)
-            async for row in query_result:
-                result.update(set(row.values()))
+            num_entries = await conn.scalar(
+                sa.select([func.count()])
+                .select_from(projects)
+                .where(projects.c.workbench.op("->>")(f"{node_id}") != None)
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/projects/projects_db.py:817]: `.op("->>")(f"{node_id}")` is this an exist check? How does it work?

**Review by GitHK — APPROVED** [review]
Very nice. One question below.

File: services/web/server/src/simcore_service_webserver/projects/projects_db.py (hunk 4 above)
Comment (sanderegg) [reply services/web/server/src/simcore_service_webserver/projects/projects_db.py:817]: You can find the postgreql calls regarding json here:
https://www.postgresql.org/docs/12/functions-json.html
This returns the value of a key in a JSON dict for example. ```!= None``` means it asks whether it is ```IS NOT NULL``` in SQL, thus yes it checks for existence.

**Pull request merged**